website/content/perf/2024-11-12-threading/index.md

```diff
@@ -0,0 +1,175 @@
+
+### Full TLS1.3 handshakes on the server
+
+This case is a common one for a TLS server, and excludes
+any need for state on the server side.
+
+![graph of TLS1.3 full handshake server performance](full-server.svg)
+
+This graph shows the OpenSSL 3 scalability problems, the improvements
+made between OpenSSL 3.0 and 3.3, and the absence of such problems
+in BoringSSL and rustls.
+
+### Resumed TLS1.2 handshakes on the server
+
+This covers both ticket-based resumption, and session-id-based resumption.
+The latter requires the server to maintain a cache across threads, so some
+"droop" is expected in these traces.
+
+![graph of TLS1.2 resumed handshake server performance](resumed-12-server.svg)
+
+Clearly something is not right with rustls's ticket resumption performance.
+
+### Resumed TLS1.3 handshakes on the server
+
+This covers only ticket-based resumption, so the server needs no state between
+threads.
+
+![graph of TLS1.3 resumed handshake server performance](resumed-13-server.svg)
+
+Again, rustls's ticket resumption performance is suspect.
+
+# Improving rustls's ticket resumption performance
+
+In rustls we have a `TicketSwitcher`, which is responsible for rotating
+between ticket keys periodically.  This is important because (especially
+in TLS1.2) compromise of the ticket key destroys the security of all past
+and future sessions.
+
+Unfortunately a mutex was held around all ticket operations -- so that
+a thread finding the key need to be rotated did not race another thread
+currently using that key.
+
+In [PR #2193](https://github.com/rustls/rustls/pull/2193) this was improved
+to use a rwlock -- this means there is no contention at all in the common case.
+
+The improvement looks like:
+
+![graph of TLS1.2 resumed handshake server performance](resumed-12-server-postfix.svg)
+
+![graph of TLS1.3 resumed handshake server performance](resumed-13-server-postfix.svg)
+
+# Measuring worse-case latency at high concurrency
+
+The above measurements only record _average handshake throughput per thread_,
+which is fine for seeing how that changes with the number of threads.
+Another important measure is: what is the latency distribution of all handshakes
+performed, at high thread counts?  What we're looking for here is evidence
+that no one handshake experiences poor performance.
+
+From here on, the versions tested move to:
+
+- openssl-bench: XXX
+- rustls: XXX
+
+In the following test, we measure handshake latency when 80 threads are working
+at once.  In the rustls test, this produces very satisfying `htop` output:
+
+![htop output, showing 80 cores at high utilization](htop-80-99.png)
+
+Note that the below charts use a base 10 logarithm x axis, to adequately show the
+four distributions in one figure.
+
+![graph of TLS1.2 resumed handshake server latency](latency-resume-tls12-server.svg)
+
+Here we can clearly see an improvement between OpenSSL 3.0 and OpenSSL 3.3.
+rustls has the tightest distribution, followed by BoringSSL.
```

Comment:
Assuming this is describing the image above, that graph shows OpenSSL 3.3 having a tighter distribution than Boring

Reply:
Because OpenSSL 3.3 is further along on the log scale, its distribution is visually narrower but covers a larger space of numbers. BoringSSL is (roughly -- reading off the graph) from 20-90us, whereas OpenSSL 3.3 is maybe 700-900us?

Reply:
Oh man, that was... not visually obvious. Obviously I should have read the axis closer though. I think a parenthetical pointing that out might help others avoid the same mistake. Suggested language:

> (OpenSSL 3.3's distribution may appear visually tighter than BoringSSL's, however the scale is logarithmic, so it is not.)